docs/RFCS/replica_tombstone.md

```diff
@@ -0,0 +1,101 @@
+- Feature Name: replica_tombstone
+- Status: draft
+- Start Date: 2015-07-29
+- RFC PR:
+- Cockroach Issue: #768
+
+# Summary
+
+Add a **Replica ID** to each replica in a range descriptor, and pass
+this replica ID in all raft messages. When a replica is removed from a
+node, record a tombstone with the replica ID, and reject any later
+messages referring to that replica.
+
+# Motivation
+
+Replica removal is necessarily an asynchronous process -- the node
+holding the removed replica may be down at the time of its removal,
+and so any recovering node may have some replicas that should have
+been removed but have not yet been garbage-collected. These nodes may
+try to send raft messages that could disrupt the legitimate members of
+the group. Even worse, if there has been enough turnover in the
+membership of a group, a quorum of removed replicas may manage to
+elect a leader among themselves.
+
+We have an additional complication compared to vanilla raft because we
+allow node IDs to be reused (this is necessary for coalesced
+heartbeats. We may remove a replica from a node and then re-add it
+later with the same node ID, so we must be able to distinguish
+messages from an earlier epoch of the range.
+
+Here is a scenario that can lead to split-brain in the current system:
+
+1. A range R has replicas on nodes A, B, and C; C is down.
+2. Nodes A and B execute a `ChangeReplicas` transactions to remove
+   node C. Several more `ChangeReplicas` transactions follow, adding
+   nodes D, E, and F and removing A and B.
+3. Nodes A and B garbage-collect their copies of the range.
+4. Node C comes back up. When it doesn't hear from the leader of
+   range R, it starts an election.
+5. Nodes A and B see that node C has a more advanced log position for
+   range R than they do (since they have nothing), so they vote for it.
+   C becomes leader and sends snapshots to A and B.
```

Comment:
A and B don't have a log position because they've garbage collected their log. Currently, they can't tell whether C is the legitimate leader of the range adding them as a replica or if it's out of date. I originally considered using the raft (term, index) in the tombstone instead of introducing replica IDs, but I'm not sure we can guarantee safety with this approach. The problem is that even if we insert a dummy write, there is no guarantee that that write will reach nodes A and B.